runtime/near-vm-logic/src/dependencies.rs

```diff
@@ -111,13 +111,37 @@ pub trait External {
     ///
     /// # let mut external = MockedExternal::new();
     /// external.storage_set(b"key42", b"value1337").unwrap();
-    /// // Returns value if exists
+    /// // Returns Ok if exists
     /// assert_eq!(external.storage_remove(b"key42"), Ok(()));
-    /// // Returns None if there was no value
+    /// // Returns Ok if there was no value
     /// assert_eq!(external.storage_remove(b"no_value_key"), Ok(()));
     /// ```
     fn storage_remove(&mut self, key: &[u8]) -> Result<()>;
 
+    /// Removes all keys under given suffix in the storage.
+    ///
+    /// # Arguments
+    ///
+    /// * `prefix` - a prefix for all keys to remove
+    ///
+    /// # Errors
+    ///
+    /// This function could return HostErrorOrStorageError::StorageError on underlying DB failure
+    ///
+    /// # Example
+    /// ```
+    /// # use near_vm_logic::mocks::mock_external::MockedExternal;
+    /// # use near_vm_logic::External;
+    ///
+    /// # let mut external = MockedExternal::new();
+    /// external.storage_set(b"key1", b"value1337").unwrap();
+    /// external.storage_set(b"key2", b"value1337").unwrap();
+    /// assert_eq!(external.storage_remove_subtree(b"key"), Ok(()));
+    /// assert!(!external.storage_has_key(b"key1").unwrap());
+    /// assert!(!external.storage_has_key(b"key2").unwrap());
+    /// ```
+    fn storage_remove_subtree(&mut self, prefix: &[u8]) -> Result<()>;
```

Comment:
We should try to avoid having this operation. It requires us to have sorted `Trie` with iterators which are problematic. The only operation you need is to remove storage under `Address`, ideally you would simulate it with random access operations instead rely on the subtree.

There are only one other place required for the protocol to have iterator and this is `DeleteAccount`. We had plans to reimplement ordered tree as `UnorderedMap` with ability of iterating over keys.

Reply:
Yes, this pretty much DeleteAccount just under subtree.
When we reimplement -> we can change this part as well.

Reply:
The abuse with `DeleteAccount` was to create a large sub-trie and later revert the action effectively iterating over the trie for the low cost. Once `DeleteAccount` become the last action, it removed this abuse, while still not accounting for the size. If `storage_remove_subtree` can be called and then reverted, then it will be possible to abuse it.